keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+            * [Add Parameter Wizard kudo package add parameter](#add-parameter-wizard-kudo-package-add-parameter)
+            * [Parameter Listing kudo package list parameters](#parameter-listing-kudo-package-list-parameters)
+      * [Additional attributes in JSON-schema](#additional-attributes-in-json-schema)
+         * [Trigger](#trigger)
+         * [Immutable](#immutable)
+         * [OldName](#oldname)
+      * [Operator conversion and upgrade example](#operator-conversion-and-upgrade-example)
+         * [kudo.dev/v1beta1 operator version](#kudodevv1beta1-operator-version)
+         * [Conversion to kudo.dev/v1beta2](#conversion-to-kudodevv1beta2)
+         * [Upgrading operator](#upgrading-operator)
+      * [Resources](#resources)
+
+<!--te-->
+
+## Summary
+
+Replace current flat "list of parameters" with a single, nested and typed structure described by [JSON-schema](https://json-schema.org/).
+
+## Motivation
+
+Currently, KUDO uses a flat [list of parameter](https://github.com/kudobuilder/kudo/blob/main/pkg/apis/kudo/v1beta1/parameter_types.go#L4) which, while being simple and accessible, has its limits like the need to use name prefixes to group the parameters. [JSON-schema](https://json-schema.org) is a de-facto standard when describing a data format. It can also be expressed as YAML to fit into the Kubernetes ecosystem. Along with naming and descriptions it provides following improvements over the existing parameter schema:   
+- Labels
+- Nested parameters
+- Validation and ranged values
+- More parameter types (enums, numbers, etc.)
+- Easily extendable with custom attributes like `trigger` and `immutable`
+- Rich ecosystem of tools and libraries e.g. [UI generators](http://json-schema.org/implementations.html#generators-from-schemas)
+
+Latter point might help to build an ecosystem around KUDO operators which might help driving KUDOs adoption.
+
+### Goals
+
+Introduce a new parameter format based on the JSON-schema.
+
+### Non-Goals
+
+Extend the current format to support new features.
+
+### Proposal Overview
+
+One part of the proposal deals the client-side of the new format. This includes:
+ - Introduce new API version (`apiVersion:kudo.dev/v1beta2`) for the operator package `params.yaml`. KUDO should be able to read both old and new formats.
+- Support updating nested values from the command line e.g. `kudo update ... -p some.nested.key=value`
+- Support updating nested values using values file e.g. `kudo update ... -P values.yaml`
+- Add nested parameters via. the wizard `kudo package add parameter`
+- Parameter listing aka `kudo package list parameters`. We need a new output format here - maybe a tree-table?
+- Possible additional validation for our custom attributes e.g. parameter `trigger`
+
+The server-side of the proposal includes:
+- Updating `Operator` and `Instance` CRDs to use JSON-schema as the parameter format  
+- Providing a conversion-webhook to convert in-cluster running operators to the new format 
```

Comment:
just a reminder that this will bump minimal version for kudo to 1.16 while we've been keeping it 1.15 (is konvoy 1.16 out?). I don't think it changes everything just noting :)

Reply:
Well, it's feature gated and on by default for 1.15. But yes, we should probably mention it somewhere in the KEP. (https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion)